File: backend/app/schemas/article.py
"""
文章 Schema 模块

本模块定义了文章相关的 Pydantic 模型，用于请求验证和响应序列化。
"""
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field
from app.schemas.user import UserBrief


class ArticleBase(BaseModel):
    """
    文章基础 Schema
    
    定义文章共有的字段。
    """
    headline: str = Field(..., min_length=1, max_length=100, description="文章标题")
    type: int = Field(..., ge=1, description="文章分类ID")
    content: Optional[str] = Field(None, description="文章内容")
    thumbnail: Optional[str] = Field(None, description="缩略图路径")
    credit: int = Field(default=0, ge=0, description="阅读所需积分")


class ArticleCreate(ArticleBase):
    """
    创建文章 Schema
    
    用于创建文章时的请求验证。
    """
    drafted: int = Field(default=0, ge=0, le=1, description="是否草稿 (0:否, 1:是)")


class ArticleUpdate(BaseModel):
    """
    更新文章 Schema
    
    用于更新文章时的请求验证。所有字段均为可选。
    """
    headline: Optional[str] = Field(None, min_length=1, max_length=100, description="文章标题")
    type: Optional[int] = Field(None, ge=1, description="文章分类ID")
    content: Optional[str] = Field(None, description="文章内容")
    thumbnail: Optional[str] = Field(None, description="缩略图路径")
    credit: Optional[int] = Field(None, ge=0, description="阅读所需积分")
    drafted: Optional[int] = Field(None, ge=0, le=1, description="是否草稿")


class ArticleResponse(BaseModel):
    """
    文章响应 Schema
    
    用于返回文章详情。
    """
    articleid: int = Field(..., description="文章ID")
    userid: int = Field(..., description="作者ID")
    type: int = Field(..., description="文章分类ID")
    headline: str = Field(..., description="文章标题")
    content: Optional[str] = Field(None, description="文章内容")
    thumbnail: Optional[str] = Field(None, description="缩略图路径")
    credit: int = Field(..., description="阅读所需积分")
    readcount: int = Field(..., description="阅读次数")
    replycount: int = Field(..., description="回复次数")
    recommended: int = Field(..., description="是否推荐")
    hidden: int = Field(..., description="是否隐藏")
    drafted: int = Field(..., description="是否草稿")
    checked: int = Field(..., description="是否审核通过")
    createtime: Optional[datetime] = Field(None, description="创建时间")
    updatetime: Optional[datetime] = Field(None, description="更新时间")
    author: Optional[UserBrief] = Field(None, description="作者信息")
    
    class Config:
        from_attributes = True


class ArticleListItem(BaseModel):
    """
    文章列表项 Schema
    
    用于文章列表接口返回，包含文章的简要信息。
    """
    articleid: int = Field(..., description="文章ID")
    userid: int = Field(..., description="作者ID")
    type: int = Field(..., description="文章分类ID")
    headline: str = Field(..., description="文章标题")
    content: Optional[str] = Field(None, description="文章摘要")
    thumbnail: Optional[str] = Field(None, description="缩略图路径")
    credit: int = Field(..., description="阅读所需积分")
    readcount: int = Field(..., description="阅读次数")
    replycount: int = Field(..., description="回复次数")
    recommended: int = Field(..., description="是否推荐")
    createtime: Optional[datetime] = Field(None, description="创建时间")
    author: Optional[UserBrief] = Field(None, description="作者信息")
    
    class Config:
        from_attributes = True


class ArticleTypeConfig(BaseModel):
    """
    文章类型配置 Schema
    
    用于返回文章分类配置。
    """
    types: dict = Field(..., description="文章分类字典")

File: backend_cpp/models/Article.cc
/**
 * @file Article.cc
 * @brief Article Model Implementation
 */

#include "Article.h"
#include "core/logger.h"

namespace woniunote {
namespace models {

Article::Article(const drogon::orm::Row& row)
{
    if (!row["articleid"].isNull()) articleid_ = row["articleid"].as<int64_t>();
    Logger::debug("[Model] Article loaded", {{"articleid", std::to_string(articleid_)}});
    if (!row["userid"].isNull()) userid_ = row["userid"].as<int64_t>();
    if (!row["type"].isNull()) type_ = row["type"].as<int>();
    if (!row["headline"].isNull()) headline_ = row["headline"].as<std::string>();
    if (!row["content"].isNull()) content_ = row["content"].as<std::string>();
    if (!row["thumbnail"].isNull()) thumbnail_ = row["thumbnail"].as<std::string>();
    if (!row["credit"].isNull()) credit_ = row["credit"].as<int>();
    if (!row["readcount"].isNull()) readcount_ = row["readcount"].as<int>();
    if (!row["replycount"].isNull()) replycount_ = row["replycount"].as<int>();
    if (!row["recommended"].isNull()) recommended_ = row["recommended"].as<int>();
    if (!row["hidden"].isNull()) hidden_ = row["hidden"].as<int>();
    if (!row["drafted"].isNull()) drafted_ = row["drafted"].as<int>();
    if (!row["checked"].isNull()) checked_ = row["checked"].as<int>();
    if (!row["createtime"].isNull()) createtime_ = row["createtime"].as<std::string>();
    if (!row["updatetime"].isNull()) updatetime_ = row["updatetime"].as<std::string>();
    // Join field from users table
    if (!row["nickname"].isNull()) nickname_ = row["nickname"].as<std::string>();
}

// Helper function to generate stable thumbnail filename based on article type
// Returns just the filename (e.g., "906.png"), NOT the full path
// Frontend will prepend "/api/thumb/" prefix
static std::string getStableThumbnail(const std::string& thumbnail, int type)
{
    // If thumbnail is already set and valid (just filename), use it
    if (!thumbnail.empty() && thumbnail.find(".png") != std::string::npos) {
        // If it's a full path starting with /api/thumb/, extract just the filename
        size_t pos = thumbnail.rfind('/');
        if (pos != std::string::npos) {
            return thumbnail.substr(pos + 1);
        }
        return thumbnail;
    }
    // Generate thumbnail filename based on article type (e.g., "906.png")
    return std::to_string(type) + ".png";
}

Json::Value Article::toJson() const
{
    Logger::debug("[Model] Article toJson", {{"articleid", std::to_string(articleid_)}, {"headline", headline_}});
    Json::Value ret;
    ret["articleid"] = static_cast<Json::Int64>(articleid_);
    ret["userid"] = static_cast<Json::Int64>(userid_);
    ret["type"] = type_;
    ret["headline"] = headline_;
    ret["content"] = content_;
    ret["thumbnail"] = getStableThumbnail(thumbnail_, type_);
    ret["credit"] = credit_;
    ret["readcount"] = readcount_;
    ret["replycount"] = replycount_;
    ret["recommended"] = recommended_;
    ret["hidden"] = hidden_;
    ret["drafted"] = drafted_;
    ret["checked"] = checked_;
    ret["createtime"] = createtime_;
    ret["updatetime"] = updatetime_;
    // Wrap author info in author object for frontend compatibility
    Json::Value author;
    author["nickname"] = nickname_;
    author["userid"] = static_cast<Json::Int64>(userid_);
    ret["author"] = author;
    return ret;
}

Json::Value Article::toJsonBrief() const
{
    Json::Value ret;
    ret["articleid"] = static_cast<Json::Int64>(articleid_);
    ret["userid"] = static_cast<Json::Int64>(userid_);
    ret["type"] = type_;
    ret["headline"] = headline_;
    ret["thumbnail"] = getStableThumbnail(thumbnail_, type_);
    ret["credit"] = credit_;
    ret["readcount"] = readcount_;
    ret["replycount"] = replycount_;
    ret["recommended"] = recommended_;
    ret["createtime"] = createtime_;
    ret["content"] = content_;  // For excerpt generation
    // Wrap author info in author object for frontend compatibility
    Json::Value author;
    author["nickname"] = nickname_;
    author["userid"] = static_cast<Json::Int64>(userid_);
    ret["author"] = author;
    return ret;
}

} // namespace models
} // namespace woniunote
